test/authentication/testExternalBrowser.js

```diff
@@ -0,0 +1,190 @@
+      assert.strictEqual(error?.message, 'Error while getting SAML token: Browser action timed out after 100 ms.');
+      await verifyConnectionIsNotUp(connection);
+    });
+  });
+
+  describe('ID Token authentication tests', async () => {
+    const connectionOption = { ...connParameters.externalBrowser, clientStoreTemporaryCredential: true };
+    const key = Util.buildCredentialCacheKey(connectionOption.host, connectionOption.username, 'ID_TOKEN');
+    const defaultCredentialManager = new JsonCredentialManager();
+    let firstIdToken;
+
+    before(async () => {
+      await defaultCredentialManager.remove(key);
+      await cleanBrowserProcesses();
+    });
+
+    afterEach(async () => {
+      await cleanBrowserProcesses();
+      await destroyConnection(connection);
+    });
+
+    it('obtains the id token from the server and saves it on the local storage', async function () {
+      connection = snowflake.createConnection(connectionOption);
+      const provideCredentialsPromise = execWithTimeout('node', [provideBrowserCredentialsPath, 'success', login, password], 15000);
+      await Promise.allSettled([connection.connectAsync(connectAsyncCallback()), provideCredentialsPromise]);
+      verifyNoErrorWasThrown();
+      await verifyConnectionIsUp(connection);
+    });
+
+    it('the token is saved in the credential manager', async function () {
+      firstIdToken = await defaultCredentialManager.read(key);
+      assert.notStrictEqual(firstIdToken, null);
+    });
+
+    it('authenticates by token, browser credentials not needed',  async function () {
+      connection = snowflake.createConnection(connectionOption);
+      await connection.connectAsync(connectAsyncCallback());
+      verifyNoErrorWasThrown();
+      await verifyConnectionIsUp(connection);
+    });
+
+    it('opens browser okta authentication again when token is incorrect',  async function () {
+      await defaultCredentialManager.write(key, '1234');
+      connection = snowflake.createConnection(connectionOption);
+      const provideCredentialsPromise = execWithTimeout('node', [provideBrowserCredentialsPath, 'success', login, password], 15000);
+      await Promise.allSettled([connection.connectAsync(connectAsyncCallback()), provideCredentialsPromise]);
+      verifyNoErrorWasThrown();
+      await verifyConnectionIsUp(connection);
+    });
+
+    it('refreshes the token for credential cache key', async function () {
+      const newToken = await defaultCredentialManager.read(key);
+      assert.notStrictEqual(firstIdToken, newToken);
+    });
+  });
+
+  function connectAsyncCallback() {
+    return function (err) {
+      error = err;
+    };
+  }
+
+  function verifyNoErrorWasThrown() {
+    assert.equal(error, null);
+  }
+});
+
+async function cleanBrowserProcesses() {
+  exec('pkill -f chromium');
+  exec('pkill -f xdg-open');
+}
+
+async function verifyConnectionIsUp(connection) {
+  assert.ok(connection.isUp(), 'Connection is not up');
```

Comment:
We can skip verifying whether the connection is up. This is controlled inside the isValidAsync method.

Reply:
done